### ISSUE TITLE
models.lda_model.top_topics missing imports, needs tests

### ISSUE THREAD

**Issue by scrappythekangaroo** [RaRe-Technologies/gensim#323]
Inside the function definition for top_topics you will find that 
bestn = np.argsort(topic)[::-1][:num_words]
should be:
bestn = numpy.argsort(topic)[::-1][:num_words]

Likewise, lower down ifilter is being used, but has not been imported from itertools:
if len(list(ifilter(lambda x: x[0] == id,corpus[document]))) > 0:

**Comment by piskvorky** [comment]
Hmm, you're right. Thanks for reporting.

@jtmcmc how did you test this? Can you add unit tests?

**Comment by scrappythekangaroo** [comment]
Actually I think there might be more problems with top_topics.  I'll take a closer look and will get back.

**Comment by cscorley** [comment]
Hm, I compared the output to Mallet's implementation and the function acted roughly the same on the same input data, but maybe my IPython instance had some extra stuff going on in the test session I was using.

**Comment by cscorley** [comment]
I was trying to write a simple unit test based on the scores MALLET gives for its implementation. Here's the output of running our small test corpus with MALLET with a symmetric alpha & beta of 0.5:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<model>
<topic id='0' tokens='18.0000' document_entropy='1.8310' word-length='2.2500' coherence='-117.4410' uniform_dist='0.4708' corpus_dist='0.3708' eff_num_words='7.0435' token-doc-diff='0.0016' rank_1_docs='0.8571' allocation_ratio='0.8571' allocation_count='1.0000'>
<word rank='1' count='4' prob='0.22222' cumulative='0.22222' docs='3' word-length='6.0000' coherence='0.0000' uniform_dist='0.2180' corpus_dist='0.1060' token-doc-diff='0.0013'>system</word>
<word rank='2' count='3' prob='0.16667' cumulative='0.38889' docs='3' word-length='5.0000' coherence='-1.9459' uniform_dist='0.1155' corpus_dist='0.0795' token-doc-diff='0.0001'>trees</word>
<word rank='3' count='2' prob='0.11111' cumulative='0.50000' docs='2' word-length='3.0000' coherence='-1.9459' uniform_dist='0.0320' corpus_dist='0.0530' token-doc-diff='0.0000'>eps</word>
<word rank='4' count='2' prob='0.11111' cumulative='0.61111' docs='2' word-length='4.0000' coherence='-1.9459' uniform_dist='0.0320' corpus_dist='0.0079' token-doc-diff='0.0000'>user</word>
<word rank='5' count='2' prob='0.11111' cumulative='0.72222' docs='2' word-length='9.0000' coherence='-1.9459' uniform_dist='0.0320' corpus_dist='0.0530' token-doc-diff='0.0000'>interface</word>
<word rank='6' count='2' prob='0.11111' cumulative='0.83333' docs='2' word-length='5.0000' coherence='-1.9459' uniform_dist='0.0320' corpus_dist='0.0530' token-doc-diff='0.0000'>human</word>
<word rank='7' count='2' prob='0.11111' cumulative='0.94444' docs='2' word-length='8.0000' coherence='-1.9459' uniform_dist='0.0320' corpus_dist='0.0530' token-doc-diff='0.0000'>computer</word>
<word rank='8' count='1' prob='0.05556' cumulative='1.00000' docs='1' word-length='5.0000' coherence='-1.9459' uniform_dist='-0.0225' corpus_dist='-0.0345' token-doc-diff='0.0000'>graph</word>
</topic>
<topic id='1' tokens='11.0000' document_entropy='1.3730' word-length='1.6500' coherence='-142.3722' uniform_dist='0.7171' corpus_dist='0.7958' eff_num_words='5.7619' token-doc-diff='0.0000' rank_1_docs='0.7500' allocation_ratio='1.0000' allocation_count='1.0000'>
<word rank='1' count='2' prob='0.18182' cumulative='0.18182' docs='2' word-length='6.0000' coherence='0.0000' uniform_dist='0.1418' corpus_dist='0.1763' token-doc-diff='0.0000'>minors</word>
<word rank='2' count='2' prob='0.18182' cumulative='0.36364' docs='2' word-length='5.0000' coherence='0.0000' uniform_dist='0.1418' corpus_dist='0.1025' token-doc-diff='0.0000'>graph</word>
<word rank='3' count='2' prob='0.18182' cumulative='0.54545' docs='2' word-length='4.0000' coherence='-1.6094' uniform_dist='0.1418' corpus_dist='0.1763' token-doc-diff='0.0000'>time</word>
<word rank='4' count='2' prob='0.18182' cumulative='0.72727' docs='2' word-length='6.0000' coherence='-0.5108' uniform_dist='0.1418' corpus_dist='0.1763' token-doc-diff='0.0000'>survey</word>
<word rank='5' count='2' prob='0.18182' cumulative='0.90909' docs='2' word-length='8.0000' coherence='-1.6094' uniform_dist='0.1418' corpus_dist='0.1763' token-doc-diff='0.0000'>response</word>
<word rank='6' count='1' prob='0.09091' cumulative='1.00000' docs='1' word-length='4.0000' coherence='-1.6094' uniform_dist='0.0079' corpus_dist='-0.0117' token-doc-diff='0.0000'>user</word>
</topic>
</model>
```
while the output for the current implementation is now:

```python
[([(0.17335630427482274, u'system'),
 (0.13471218853119102, u'user'),
 (0.096139465884184067, u'eps'),
 (0.09609735674170379, u'time'),
 (0.09609735674170379, u'response'),
 (0.096094134603670237, u'computer'),
 (0.096068707755876045, u'human'),
 (0.096067848363866573, u'interface'), 
 (0.056841786554836356, u'survey'),
 (0.019576526138850872, u'trees'),
 (0.019485228292470218, u'graph'),
 (0.019463096116824254, u'minors')], -46.977660274659229)]
```

This seems to have changed when I last tested it. Why is only one topic now returned out of the two? Why is the coherence score this different? The score was within 20 points last time I tried. Do you think it was the removal of the generators, @jtmcmc?

**Comment by cscorley** [comment]
@scrappythekangaroo,

I've made some changes to the method on the `develop` branch. I refactored the method a bit and commented the code. I found a couple of typos while refactoring, and fixed them to more closely match  Eq 1 in [the paper](http://people.cs.umass.edu/~wallach/publications/mimno11optimizing.pdf). Let me know if it fixes any of the other issues you've noticed. I think there still may be something with the `num_words` limitation; it doesn't seem to be a part of the Mallet implementation as far as I can tell.

**Comment by scrappythekangaroo** [comment]
Hi thanks a lot for the updates.  I've only just started working with gensim, so I may not be that much help just yet in making new code suggestions, but might still be able to help with the easy stuff.   I did notice that your code in the develop branch is missing the num_topics from the final top_topics list.

**Comment by piskvorky** [comment]
Sorry this was your first experience with gensim @scrappythekangaroo !

You came across new functionality that wasn't well tested -- this is not "standard" mode.

Thanks for investigating & fixing @cscorley .

**Comment by scrappythekangaroo** [comment]
No worries, I'm finding it a very useful package!

**Comment by cscorley** [comment]
@scrappythekangaroo I've removed the "num_topics" parameter because the scoring calculates every topic's score anyway, so there is no reason to limit the data returned when the computation has already been done. Whereas the other methods with the num_topics, such as `show_topics`, take a random pick and truly only use that many topics. To have only the top 10 topics, a user can instead do `model.top_topics(corpus)[:10]` themselves.

**Comment by menshikh-iv** [comment]
Resolved in #4
